cdap-ui/app/cdap/components/DataPrepConnections/SpannerConnection/index.tsx

```diff
@@ -0,0 +1,382 @@
+        <button
+          className="btn btn-primary"
+          onClick={onClickFn}
+          disabled={disabled}
+        >
+          {T.translate(`${PREFIX}.Buttons.${this.props.mode}`)}
+        </button>
+
+        {this.renderTestButton()}
+      </ModalFooter>
+    );
+  }
+
+  private renderContent() {
+    if (this.state.loading) {
+      return (
+        <div className="spanner-detail text-xs-center">
+          <br />
+          <LoadingSVG />
+        </div>
+      );
+    }
+
+    return (
+      <div className="spanner-detail">
+
+        <div className="form">
+          <div className="form-group row">
+            <label className={LABEL_COL_CLASS}>
+              {T.translate(`${PREFIX}.name`)}
+              <span className="asterisk">*</span>
+            </label>
+            <div className={INPUT_COL_CLASS}>
+              <div className="input-text">
+                <input
+                  type="text"
+                  className="form-control"
+                  value={this.state.name}
+                  onChange={this.handleChange.bind(this, 'name')}
+                  disabled={this.props.mode === 'EDIT'}
+                />
+              </div>
+            </div>
+          </div>
+
+          <div className="form-group row">
+            <label className={LABEL_COL_CLASS}>
+              {T.translate(`${PREFIX}.projectId`)}
+              <span className="asterisk">*</span>
+            </label>
+            <div className={INPUT_COL_CLASS}>
+              <div className="input-text">
+                <input
+                  type="text"
+                  className="form-control"
+                  value={this.state.projectId}
+                  onChange={this.handleChange.bind(this, 'projectId')}
+                />
+              </div>
+            </div>
+          </div>
+
+          <div className="form-group row">
+            <label className={LABEL_COL_CLASS}>
+              {T.translate(`${PREFIX}.serviceAccountKeyfile`)}
+              <span className="asterisk">*</span>
+            </label>
+            <div className={INPUT_COL_CLASS}>
+              <div className="input-text">
+                <input
+                  type="text"
+                  className="form-control"
+                  value={this.state.serviceAccountKeyfile}
+                  onChange={this.handleChange.bind(this, 'serviceAccountKeyfile')}
+                />
+              </div>
+            </div>
+          </div>
+
+        </div>
+      </div>
+    );
+  }
+
+  private renderMessage() {
+    if (!this.state.error && !this.state.connectionResult.message) { return null; }
+
+    if (this.state.error) {
+      return (
+        <CardActionFeedback
+          type={this.state.connectionResult.type}
+          message={T.translate(`${PREFIX}.ErrorMessages.${this.props.mode}`)}
+          extendedMessage={this.state.error}
+        />
+      );
+    }
+
+    const connectionResult = this.state.connectionResult;
```

Comment:
nit: Can we move this to the top of the function and reuse the result type and extended message in both the return statements?